File: frontend/src/features/products/services/categoryService.ts
import apiClient from '@/services/api/apiClient';
import { CategoryResponse, CategoryRequest } from '../types/product';
import { ApiResponse } from '@/types/apiResponse';

const CATEGORY_API = '/v1/categories';

export const categoryService = {
  /**
   * Get full category tree structure
   */
  async getCategoryTree(): Promise<CategoryResponse[]> {
    const response = await apiClient.get<ApiResponse<CategoryResponse[]>>(CATEGORY_API);
    return response.data.data || [];
  },

  /**
   * Get category by ID or slug
   */
  async getCategory(idOrSlug: string | number): Promise<CategoryResponse> {
    const response = await apiClient.get<ApiResponse<CategoryResponse>>(`${CATEGORY_API}/${idOrSlug}`);
    return response.data.data;
  },

  /**
   * Create a new category (ADMIN only)
   */
  async createCategory(data: CategoryRequest): Promise<CategoryResponse> {
    const response = await apiClient.post<ApiResponse<CategoryResponse>>(CATEGORY_API, data);
    return response.data.data;
  },

  /**
   * Update an existing category (ADMIN only)
   */
  async updateCategory(categoryId: number, data: CategoryRequest): Promise<CategoryResponse> {
    const response = await apiClient.put<ApiResponse<CategoryResponse>>(`${CATEGORY_API}/${categoryId}`, data);
    return response.data.data;
  },

  /**
   * Delete an existing category (ADMIN only)
   */
  async deleteCategory(categoryId: number): Promise<void> {
    await apiClient.delete<ApiResponse<void>>(`${CATEGORY_API}/${categoryId}`);
  }
};

File: backend/src/main/java/com/nexoracommerce/category/service/CategoryService.java
package com.nexoracommerce.category.service;

import com.nexoracommerce.category.dto.request.CategoryRequest;
import com.nexoracommerce.category.dto.response.CategoryResponse;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.util.List;

public interface CategoryService {
    /**
     * Lấy danh sách danh mục dưới dạng cây (tree)
     */
    List<CategoryResponse> getCategoryTree();

    /**
     * Lấy chi tiết danh mục theo ID hoặc Slug
     */
    CategoryResponse getCategory(String idOrSlug);

    /**
     * Lấy chi tiết danh mục theo ID
     */
    CategoryResponse getCategoryById(Long id);

    /**
     * Lấy chi tiết danh mục theo Slug
     */
    CategoryResponse getCategoryBySlug(String slug);

    /**
     * Tạo danh mục mới (Admin)
     */
    CategoryResponse createCategory(CategoryRequest request);

    /**
     * Cập nhật danh mục (Admin)
     */
    CategoryResponse updateCategory(Long id, CategoryRequest request);

    /**
     * Xóa danh mục (Admin)
     */
    void deleteCategory(Long id);

    /**
     * Lấy danh sách danh mục có phân trang
     */
    Page<CategoryResponse> getCategoriesPageable(Pageable pageable);
}
